src/app/multi-line-chart/multi-line-chart.component.ts

```diff
@@ -34,68 +39,50 @@ export class MultiLineChartComponent implements OnInit {
     const width = 800;
     const height = 600;
 
-    const margin = { top: 20, right: 20, bottom: 30, left: 30 };
+    const margin = { top: 20, right: 20, bottom: 30, left: 60 };
 
     const data = {
       y: this.yAxisLabel,
       x: this.xAxisLabel,
       series: this.data,
       dates: times(Math.max(...this.data.map(v => v.values.length)), Number),
     };
-    console.log(data);
 
-    const y = d3
+    this.yScale = d3
       .scaleLog()
-      .domain([2, d3.max(data.series, d => d3.max(d.values))])
-      .nice()
+      .domain(
+        this.yAxisBounds || [1, d3.max(data.series, d => d3.max(d.values))]
+      )
       .range([height - margin.bottom, margin.top]);
 
-    const x = d3
+    this.xScale = d3
       .scaleLinear()
-
-      .domain(d3.extent(data.dates as Number[]))
+      .domain(this.xAxisBounds || d3.extent(data.dates as Number[]))
       .range([margin.left, width - margin.right]);
 
+    console.log(this.xScale.domain(), this.yScale.domain());
+
     const xAxis = g =>
-      g
-        .attr('transform', `translate(0,${height - margin.bottom})`)
-        .call(
-          d3
-            .axisBottom(x)
-            .ticks(width / 80)
-            .tickSizeOuter(0)
-        )
-        .call(g =>
-          g
-            .select('.tick:last-of-type text')
-            .clone()
-            .attr('y', 30)
-            .attr('x', -70)
-            .attr('text-anchor', 'start')
-            .attr('font-weight', 'bold')
-            .text(data.x)
-        );
+      g.attr('transform', `translate(0,${height - margin.bottom})`).call(
+        d3
+          .axisBottom(this.xScale)
+          .tickValues(this.getDayTicks())
+          .tickSizeOuter(0)
+      );
 
     const yAxis = g =>
-      g
-        .attr('transform', `translate(${margin.left},0)`)
-        .call(d3.axisLeft(y))
-        .call(g => g.select('.domain').remove())
-        .call(g =>
-          g
-            .select('.tick:last-of-type text')
-            .clone()
-            .attr('x', 10)
```

Comment:
These attr values -- "x" and "y" can be set to adjust the positioning of the axes. I think the warrant keeping -- more info is better. You can shift the text up maybe 20 or so pixels ("y", -20) and it should be readable without clashing with the gridline

Reply:
ah i'll try it out! 👍